### PR TITLE
Finish BASP implementation

### PR DESCRIPTION
**Pull request by Neverlord**
This is a big PR again, but it brings BASP in a usable state and re-implements synchronization of endpoint managers with the multiplexer (the previous design was flawed and introduced subtle race conditions).

### CODE REVIEW

File: libcaf_net/caf/net/endpoint_manager_impl.hpp [hunk 1]
```diff
@@ -87,25 +89,36 @@ class endpoint_manager_impl : public endpoint_manager {
...
+            });
+          visit(f, ptr->value);
         }
-      } while (!this->events_.try_block());
+      } while (!q.empty());
     }
-    return transport_.handle_write_event(*this);
+    if (!transport_.handle_write_event(*this)) {
+      if (this->queue_.blocked())
+        return false;
```
Comment (josephnoir) [inline libcaf_net/caf/net/endpoint_manager_impl.hpp]: I'm simply not entirely sure what's happening here. Would you mind summarizing it in a few words. I mean, the manager somehow dequeues events and handles them before passing the event to its transport, but I'm a bit unsure what the deficit + for loop does. Does this query the event queue before letting the transport handle the message queue or ...?

File: libcaf_net/src/ec.cpp [hunk 2]
```diff
@@ -42,6 +42,7 @@ string_view ec_names[] = {
   "unimplemented",
   "app_identifiers_mismatch",
   "invalid_payload",
+  "invalid_scheme",
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_net/src/ec.cpp]: Yeah... let's not write `to_string` from hand any more. I've written a small tool for auto-generating this stuff in the future: https://github.com/actor-framework/actor-framework/pull/927. When we got this set up in the main repository then I'll update the incubator as well.

File: libcaf_net/src/actor_proxy_impl.cpp [hunk 3]
```diff
@@ -27,7 +27,8 @@ namespace net {
 
 actor_proxy_impl::actor_proxy_impl(actor_config& cfg, endpoint_manager_ptr dst)
   : super(cfg), sf_(dst->serialize_fun()), dst_(std::move(dst)) {
-  // nop
+  CAF_ASSERT(dst_ != nullptr);
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_net/src/actor_proxy_impl.cpp]: > Making this more of a documentation thing than a realistic check.

😂

Shall we leave it for "documenting preconditions purposes"?

File: libcaf_net/caf/net/endpoint_manager_impl.hpp (hunk 1 above)
Comment (Neverlord) [reply libcaf_net/caf/net/endpoint_manager_impl.hpp]: The intrusive CAF queues implement [deficit round robin](https://en.wikipedia.org/wiki/Deficit_round_robin). By setting `q.inc_deficit(q.total_task_size());` before entering the loop we make sure that we can consume the entire thing at once.

That's of course hacky, but it's not meant as the strategy we finally want to have. Eventually, we can make use of DRR to only fetch messages as long as we have some space on the socket (maybe plus some extra buffering) for writing.

File: libcaf_net/src/actor_proxy_impl.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (josephnoir) [reply libcaf_net/src/actor_proxy_impl.cpp]: Sure, why not.

File: libcaf_net/test/endpoint_manager.cpp [hunk 4]
```diff
@@ -46,12 +45,14 @@ string_view hello_test{"hello test!"};
 struct fixture : test_coordinator_fixture<>, host_fixture {
   fixture() {
     mpx = std::make_shared<multiplexer>();
+    mpx->set_thread_id();
     if (auto err = mpx->init())
       CAF_FAIL("mpx->init failed: " << sys.render(err));
```
Comment (jakobod) [inline libcaf_net/test/endpoint_manager.cpp]: Isn't error printable without `sys.render()`?

File: libcaf_net/test/stream_application.cpp [hunk 5]
```diff
@@ -20,18 +20,21 @@
...
+#include "caf/net/test/host_fixture.hpp"
 #include "caf/test/dsl.hpp"
 
-#include "host_fixture.hpp"
-
 #include <vector>
 
 #include "caf/byte.hpp"
 #include "caf/forwarding_actor_proxy.hpp"
+#include "caf/net/actor_proxy_impl.hpp"
```
Comment (jakobod) [inline libcaf_net/test/stream_application.cpp]: Unused.

File: libcaf_net/test/stream_application.cpp [hunk 6]
```diff
@@ -20,18 +20,21 @@
...
+#include "caf/net/backend/test.hpp"
 #include "caf/net/basp/connection_state.hpp"
 #include "caf/net/basp/constants.hpp"
 #include "caf/net/basp/ec.hpp"
 #include "caf/net/make_endpoint_manager.hpp"
+#include "caf/net/middleman.hpp"
+#include "caf/net/middleman_backend.hpp"
```
Comment (jakobod) [inline libcaf_net/test/stream_application.cpp]: Unused.

Comment (jakobod) [reply libcaf_net/test/stream_application.cpp]: `#include "caf/none.hpp"` further down too.

File: libcaf_net/test/stream_transport.cpp [hunk 7]
```diff
@@ -74,18 +73,19 @@ class dummy_application {
...
-  void handle_data(Parent&, span<const byte> data) {
+  error handle_data(Parent&, span<const byte> data) {
     rec_buf_->clear();
     rec_buf_->insert(rec_buf_->begin(), data.begin(), data.end());
+    return none;
   }
 
   template <class Parent>
-  void resolve(Parent& parent, const std::string& path, actor listener) {
+  void resolve(Parent& parent, string_view path, actor listener) {
```
Comment (jakobod) [inline libcaf_net/test/stream_transport.cpp]: In `endpoint_manager.cpp` test you used `const actor& ...`.

File: libcaf_net/caf/net/backend/test.hpp [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+  error init() override;
+
+  endpoint_manager_ptr peer(const node_id& id) override;
+
+  void resolve(const uri& locator, const actor& listener) override;
+
+  strong_actor_ptr make_proxy(node_id nid, actor_id aid) override;
+
+  void set_last_hop(node_id*) override;
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_net/caf/net/backend/test.hpp]: It's an override, so I think not repeating the name here is ok.

**Pull request merged**